core/src/main/java/io/questdb/mp/Worker.java

```diff
@@ -24,121 +24,157 @@
+                    try {
+                        Job job = jobs.get(i);
+                        if (job instanceof EagerThreadSetup) {
+                            ((EagerThreadSetup) job).setup();
+                        }
+                    } finally {
+                        Unsafe.getUnsafe().storeFence();
                     }
                 }
-                setupJobs();
-                int n = jobs.size();
-                long uselessCounter = 0;
-                while (running.get() == 1) {
-                    boolean useful = false;
-                    for (int i = 0; i < n; i++) {
+
+                // enter main loop
+                long ticker = 0L;
+                while (lifecycle.get() == Lifecycle.RUNNING) {
+                    boolean runAsap = false;
+                    for (int i = 0, n = jobs.size(); i < n; i++) {
+                        jobStartMicros.set(CLOCK_MICROS.getTicks());
                         Unsafe.getUnsafe().loadFence();
                         try {
+                            runAsap |= jobs.get(i).run(workerId, runStatus);
+                        } catch (Throwable e) {
                             try {
-                                useful |= jobs.get(i).run(workerId, runStatus);
-                            } catch (Throwable e) {
-                                onError(i, e);
+                                metrics.health().incrementUnhandledErrors();
+                            } catch (Throwable t) {
+                                stdErrCritical(t);
+                            }
+                            if (log != null) {
+                                log.critical().$("unhandled error [job=").$(jobs.get(i).toString()).$(", ex=").$(e).I$();
+                            } else {
+                                stdErrCritical(e); // log regardless
+                            }
+                            if (haltOnError) {
+                                throw e;
                             }
                         } finally {
                             Unsafe.getUnsafe().storeFence();
                         }
                     }
 
-                    if (useful) {
-                        uselessCounter = 0;
+                    if (runAsap) {
+                        ticker = 0L;
                         continue;
                     }
-
-                    uselessCounter++;
-
-                    if (uselessCounter < 0) {
-                        // deal with overflow
-                        uselessCounter = sleepThreshold + 1;
+                    if (++ticker < 0L) {
+                        ticker = sleepThreshold + 1L; // overflow
                     }
-
-                    if (uselessCounter > sleepThreshold) {
+                    if (ticker > sleepThreshold) {
                         Os.sleep(sleepMs);
-                    } else if (uselessCounter > yieldThreshold) {
+                    } else if (ticker % yieldThreshold == 0L) {
```

Comment:
we should avoid division here

Reply:
would it not then yield too much? the idea in my mind is to yield every yieldThreshold until we hit the sleep threshold. If I remove the check won't we be yielding on every run until we hit the sleep threshold?

Reply:
when we reach these thresholds the urgency to run jobs is not there; the goal of these things is to achieve low CPU utilization when system is inactive. % devision is relatively expensive computationally and should be avoided where possible

Reply:
ah!